### ISSUE TITLE
Add function that retrieves a batch from a DataLoader

### ISSUE THREAD

**Issue by cowwoc** [pytorch/pytorch#65216]
## Motivation
`DataLoader.__getitem__(self, index)` retrieves a single item from a dataloader at a time. Depending on where the data is coming from, there is potentially a huge performance gain to be had if we retrieve multiple items at a time. For example, if data is coming from a database, querying multiple items at a time is much faster than running a separate query for each item.

## Pitch

Add `DataLoader.get_batch(self, indexes) -> Iterable[Tuple[Tensor, Tensor]]` (use whatever name or method signature you like) whose default implementation delegates to `DataLoader.__getitem__(self, index)`. This would retain backwards compatibility for existing implementations while allowing users to override the batch operation to improve performance.

This would reduce the need for worker processes in some cases and reduce overall memory usage.

**Comment by ejguan** [comment]
I think it's still possible by using a custom Sampler to yield multiple indices per iteration. And you custom Dataset should be able to get batch of elements using the batch of indices.

I think this is also the reason we are working on DataLoaderV2 to push all data manipulation operations like batching and shuffling out of the responsibility of DataLoader.

**Comment by ptrblck** [comment]
@cowwoc `DataLoader`s don't necessarily know how the underlying `Dataset` is implemented, e.g. when using an [`IterableDataset`](https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset) no `__getitem__` method and thus also no `index` would be used.
As @ejguan explained, I believe a `BatchSampler` as described [here](https://discuss.pytorch.org/t/how-to-use-batchsampler-with-getitem-dataset/78788/4?u=ptrblck) could work instead.

**Comment by cowwoc** [comment]
@ptrblck Thanks for the link. Is anything expected to change for users of the `DataLoader`?

Previously `Dataset.__getitem__()` would convert a single index to a single sample. Now it will convert a batch of indexes into a batch of samples. Should users who iterate over the `DataLoader` iterator expect to see a change to the shape of data they receive or will this change be seamless to them somehow?

**Comment by cowwoc** [comment]
I found an example of the right way to do this: https://discuss.pytorch.org/t/force-dataloader-to-fetch-batched-index-from-custom-batch-sampler/86656

Counter-intuitively, you actually want to pass a `BatchSampler` into `DataLoader.sampler`.

```
class MyDataset(Dataset):
    def __init__(self):
        self.data = torch.arange(100).view(100, 1).float()
        
    def __getitem__(self, index):
        print(index)
        x = self.data[index]
        return x
    
    def __len__(self):
        return len(self.data)

dataset = MyDataset()        
sampler = torch.utils.data.sampler.BatchSampler(
    torch.utils.data.sampler.RandomSampler(dataset),
    batch_size=10,
    drop_last=False)

loader = DataLoader(
    dataset,
    sampler=sampler)

for data in loader:
    print(data)
```

> If you run it, you’ll see that the index inside the Dataset.__getitem__ will contain 10 indices, which can be used to slice the data directly.
